src/webnn/native/dml/BackendDML.cpp

```diff
@@ -0,0 +1,103 @@
+
+namespace webnn::native::dml {
+
+    Backend::Backend(InstanceBase* instance)
+        : BackendConnection(instance, wnn::BackendType::DirectML) {
+    }
+
+    MaybeError Backend::Initialize() {
+        return {};
+    }
+
+    // An adapter called the "Microsoft Basic Render Driver" is always present. This adapter is a
+    // render-only device that has no display outputs.
+    inline bool IsSoftwareAdapter(IDXGIAdapter1* pAdapter) {
+        DXGI_ADAPTER_DESC1 pDesc;
+        pAdapter->GetDesc1(&pDesc);
+        // See here for documentation on filtering WARP adapter:
+        // https://docs.microsoft.com/en-us/windows/desktop/direct3ddxgi/d3d10-graphics-programming-guide-dxgi#new-info-about-enumerating-adapters-for-windows-8
+        return pDesc.Flags == DXGI_ADAPTER_FLAG_SOFTWARE ||
+               (pDesc.VendorId == 0x1414 && pDesc.DeviceId == 0x8c);
+    }
+
+    HRESULT EnumAdapter(DXGI_GPU_PREFERENCE gpuPreference,
+                        bool useGpu,
+                        ComPtr<IDXGIAdapter1> adapter) {
+        ComPtr<IDXGIFactory6> dxgiFactory;
+        WEBNN_RETURN_IF_FAILED(CreateDXGIFactory1(IID_PPV_ARGS(&dxgiFactory)));
+        if (useGpu) {
+            UINT adapterIndex = 0;
+            while (dxgiFactory->EnumAdapterByGpuPreference(adapterIndex++, gpuPreference,
+                                                           IID_PPV_ARGS(&adapter)) !=
+                   DXGI_ERROR_NOT_FOUND) {
+                if (!IsSoftwareAdapter(adapter.Get())) {
+                    break;
+                }
+            }
+        } else {
+            WEBNN_RETURN_IF_FAILED(dxgiFactory->EnumWarpAdapter(IID_PPV_ARGS(&adapter)));
+        }
+        return S_OK;
+    }
+
+    ContextBase* Backend::CreateContext(ContextOptions const* options) {
+        wnn::DevicePreference devicePreference =
+            options == nullptr ? wnn::DevicePreference::Default : options->devicePreference;
+        bool useGpu = devicePreference == wnn::DevicePreference::Cpu ? false : true;
+        DXGI_GPU_PREFERENCE gpuPreference = DXGI_GPU_PREFERENCE_UNSPECIFIED;
+        wnn::PowerPreference powerPreference =
+            options == nullptr ? wnn::PowerPreference::Default : options->powerPreference;
+        switch (powerPreference) {
+            case wnn::PowerPreference::High_performance:
+                gpuPreference = DXGI_GPU_PREFERENCE::DXGI_GPU_PREFERENCE_HIGH_PERFORMANCE;
+                break;
+            case wnn::PowerPreference::Low_power:
+                gpuPreference = DXGI_GPU_PREFERENCE::DXGI_GPU_PREFERENCE_MINIMUM_POWER;
+                break;
+            default:
+                break;
+        }
+
+        bool useDebugLayer = false;
+#ifdef _DEBUG
+        useDebugLayer = true;
+#endif
+        ComPtr<IDXGIAdapter1> adapter;
+        if (FAILED(EnumAdapter(gpuPreference, useGpu, adapter))) {
+            dawn::ErrorLog() << "Failed to enumerate adapters.";
+            DAWN_ASSERT(0);

```

Comment:
We need to actually handle this error, not just ASSERT. Perhaps we can return `nullptr` or some error object?

Reply:
Because although enumAdapter failed, we can use nullptr(adapter) to create context, so we need not return nullptr(context).

Reply:
`DAWN_ASSERT(0)` always signals SIGABRT which crashes the process. The process shouldn't crash if EnumAdapter fails, no?

Reply:
I agree, return nullptr(context) if enumAdapter failed, is that OK?